flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobGraph.java

```diff
@@ -581,6 +582,8 @@ public String toString() {
 	}
 
 	public void uploadUserArtifacts(InetSocketAddress blobServerAddress, Configuration clientConfig) throws IOException {
```

Comment:
Could we move this method out of the `JobGraph`? It basically couples the `JobGraph` with the `BlobClient` which is not necessary. I would rather prefer that the submission method of the `JobGraph` makes sure that everything is properly uploaded and then updates the `DistributedCacheEntries`.

Reply:
This also applies to `uploadUserJars`, yes?

Reply:
Yes